src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -263,7 +263,9 @@
                                        [:= :certnames.certname :fs.certname]
 
                                        :reports
-                                       [:= :certnames.latest_report_id :reports.id]
+                                       [:and
+                                        [:= :certnames.certname :reports.certname]
+                                        [:= :certnames.latest_report_id :reports.id]]
```

Comment:
I have no idea what I was thinking before; this is not a backport. Added that condition, and will open this on stable as well.